Sources/Workspace/Workspace.swift

```diff
@@ -569,16 +569,18 @@ public class Workspace {
         reset: Bool = false
     ) throws {
         if reset {
-            try pinsStore.unpinAll()
+            pinsStore.unpinAll()
+            try pinsStore.saveState()
         }
         // Start pinning each dependency.
         for dependencyManifest in dependencyManifests.dependencies {
-            try pin(
-                pinsStore: pinsStore,
+            pin(pinsStore: pinsStore,
                 dependency: dependencyManifest.dependency,
                 package: dependencyManifest.manifest.name,
                 reason: reason)
         }
+        
+        try pinsStore.saveState()
```

Comment:
So the idea is to compute and hold state of pins store and then try to save it as a separate operation so it can be retired in case it fails. For example, client calls pinAll() and then saveState(). Client can be sure that pinAll() won't fail because it doesn't throw. Now it tries to save the state but something on the file system is making that operation fail (maybe the space ran out). Now, clients can retry this operation without calling pinAll(), recomputing everything it had already done. It is true that edit, update etc are throwing right now but we do need to refactor them at some point. 
Does this make sense?